### PR TITLE
[IMP] website_slides: allow course to have a prerequisite

### PR DESCRIPTION
**Pull request by pko-odoo**
This commit allows the course manager to add a prerequisite course to a course so that the final user will be notified on the frontend he has to take a course before the one he wants.
He could however request access to the course and take it without taking the prerequisite before if the access has been granted.
It's not possible to delete a prerequisite course without deleting the subsequent course at the same time.

Task-3213628

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_slides/views/website_slides_templates_course.xml [hunk 1]
```diff
@@ -333,6 +302,71 @@
                 </div>
             </div>
         </t>
+        <t t-elif="channel.enroll in ['public', 'invite']" name="o_wslides_prerequisite_test">
+            <div t-if="channel.prerequisite_channel_id and not channel.has_completed_prerequisite"
```
Comment (std-odoo) [inline addons/website_slides/views/website_slides_templates_course.xml:306]: shouldn' we also check that in `_action_add_members` ? (currently it's possible to join the course with a RPC call)

if you do it you can add a small test

**Review by std-odoo — COMMENTED** [review]
@pko-odoo Hi :) 

Cool feature :) I have some suggestion / question

Also I don't know if we should use a many2many instead of a many2one for prerequisite_channel_id (see messages on the task)

Cheers

File: addons/website_slides/views/website_slides_templates_course.xml (hunk 1 above)
Comment (pko-odoo) [reply addons/website_slides/views/website_slides_templates_course.xml:306]: Actually no, it's purely informative. Anyone can join a course with a prerequisite without finishing it first :smiley:

File: addons/website_slides/views/website_slides_templates_homepage.xml [hunk 2]
```diff
@@ -351,6 +351,14 @@
                         </t>
                     </t>
                 </div>
+                <div t-if="channel.prerequisite_channel_ids" class="fw-light pt-2">
+                    Prerequisite: [
+                    <span t-field="channel.prerequisite_channel_ids[0].name"/>
+                    <span t-foreach="channel.prerequisite_channel_ids[1:]" t-as="channel">
+                        , <t t-out="channel.name"/>
```
Comment (std-odoo) [inline addons/website_slides/views/website_slides_templates_homepage.xml:358]: shouldn't you do t-field to be consistent with above ?

File: addons/website_slides/models/slide_channel.py [hunk 3]
```diff
@@ -540,6 +551,24 @@ def _compute_action_rights(self):
                 channel.can_comment = user_karma >= channel.karma_slide_comment
                 channel.can_vote = user_karma >= channel.karma_slide_vote
 
+    @api.depends('visibility')
+    def _compute_prerequisite_channel_ids(self):
+        for channel in self:
+            if channel.prerequisite_channel_ids and channel.visibility != channel.prerequisite_channel_ids[0].visibility:
```
Comment (std-odoo) [inline addons/website_slides/models/slide_channel.py:557]: is it exactly what we want (only the first visibility matter) ? (better to double check with po)

**Review by std-odoo — COMMENTED** [review]
@pko-odoo Hi :) 

Last polyshing comments, otherwise LGTM :)

File: addons/website_slides/models/slide_channel.py (hunk 3 above)
Comment (pko-odoo) [reply addons/website_slides/models/slide_channel.py]: Normally the `prerequisite_channel_ids` can only be of the same type but I changed to
`if all(c.visibility != channel.visibility for c in channel.prerequisite_channel_ids):` :smiley:

File: addons/website_sale_slides/views/website_slides_templates.xml [hunk 4]
```diff
@@ -125,6 +125,21 @@
                 <span t-field="channel.product_id.list_price" t-options="{'widget': 'monetary', 'display_currency': product_info['currency_id']}"/>
             </div>
         </div>
+        <t t-if="channel.prerequisite_channel_ids and not channel.has_completed_prerequisites">
+            <small t-if="len(channel.prerequisite_channel_ids) == 1" class="text-center">
+                Prerequisite:
+                <a t-attf-href="/slides/{{slug(channel.prerequisite_channel_ids[0])}}">
+                    <span t-out="channel.prerequisite_channel_ids[0].name"/>
```
Comment (std-odoo) [inline addons/website_sale_slides/views/website_slides_templates.xml:132]: I think it can be simplify, because you already check above the length (same line just above)
```suggestion
                    <span t-out="channel.prerequisite_channel_ids.name"/>
```

File: addons/website_slides/views/website_slides_templates_course.xml [hunk 5]
```diff
@@ -333,6 +302,74 @@
...
+                    <h6>
+                        <i class="fa fa-lock"/>
+                        <span>Course Locked</span>
+                    </h6>
+                    <div>
+                        <small>
+                            Finish
+                            <a t-if="len(channel.prerequisite_channel_ids) == 1"
+                               t-attf-href="/slides/{{slug(channel.prerequisite_channel_ids[0])}}">
+                                <span t-out="channel.prerequisite_channel_ids[0].name"/>
```
Comment (std-odoo) [inline addons/website_slides/views/website_slides_templates_course.xml:318]: same here

**Review by std-odoo — COMMENTED** [review]
@pko-odoo Hi :) 

Some small comments post-validation

Cheers :)

**Comment by pko-odoo** [comment]
> 

Done ! Thanks @std-odoo :smiley:

File: addons/website_sale_slides/views/website_slides_templates.xml [hunk 6]
```diff
@@ -92,8 +92,8 @@
 
 <template name="Course Sidebar (infos, CTA)" id="course_join" inherit_id="website_slides.course_join">
     <!-- Channel main template: override button to join channel -->
-    <xpath expr="//div[hasclass('o_wslides_js_course_join')]" position="inside">
-        <t t-if="(not channel.is_member or channel.can_publish) and channel.enroll == 'payment'">
+    <xpath expr="//t[@name='o_wslides_prerequisite_test']" position="after">
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/website_sale_slides/views/website_slides_templates.xml:95]: This is rather complicated, can't we properly test the prerequisite inside this condition and keep the simple "inside" position?

Having to jump to another module to know when this is displayed is a bit confusing.

File: addons/website_sale_slides/views/website_slides_templates.xml [hunk 7]
```diff
@@ -125,6 +125,21 @@
...
+            <small t-if="len(channel.prerequisite_channel_ids) == 1" class="text-center">
+                Prerequisite:
+                <a t-attf-href="/slides/{{slug(channel.prerequisite_channel_ids)}}">
+                    <span t-out="channel.prerequisite_channel_ids.name"/>
+                </a>
+            </small>
+            <small t-else="" class="text-center">
+                There are some
+                <a href="#" class="o_wslides_js_prerequisite_course"
+                   t-att-data-channels="json.dumps([{'id': c.id, 'name': c.name} for c in channel.prerequisite_channel_ids])">
```
Comment (awa-odoo) [inline addons/website_sale_slides/views/website_slides_templates.xml:138]: Can we have "course_id" and "course_name" for the keys? A bit easier to grep.

Also "course" instead of "c".

(I think you can multiple attributes to avoid getting a huge line).

File: addons/website_slides/views/website_slides_templates_course.xml [hunk 8]
```diff
@@ -333,6 +302,74 @@
                 </div>
             </div>
         </t>
+        <t t-elif="channel.enroll in ['public', 'invite']" name="o_wslides_prerequisite_test">
+            <div t-if="channel.prerequisite_channel_ids and not channel.has_completed_prerequisites"
```
Comment (awa-odoo) [inline addons/website_slides/views/website_slides_templates_course.xml:306]: Not sure how he could have completed the prerequisites as he is not a member, but maybe I'm missing something.

File: addons/website_slides/views/website_slides_templates_course.xml [hunk 9]
```diff
@@ -333,6 +302,74 @@
...
+                    <h6>
+                        <i class="fa fa-lock"/>
+                        <span>Course Locked</span>
+                    </h6>
+                    <div>
+                        <small>
+                            Finish
+                            <a t-if="len(channel.prerequisite_channel_ids) == 1"
+                               t-attf-href="/slides/{{slug(channel.prerequisite_channel_ids)}}">
+                                <span t-out="channel.prerequisite_channel_ids.name"/>
```
Comment (awa-odoo) [inline addons/website_slides/views/website_slides_templates_course.xml:318]: Even if 100% the same I would prefer "channel.prerequisite_channel_ids[0].name".

Also I don't think ".name" is necessary, as t-outing a recordset should show its display name?

(Also maybe the t-out can be inside the a? Trying to gain some indentation levels ^^)

File: addons/website_slides/views/website_slides_templates_course.xml [hunk 10]
```diff
@@ -333,6 +302,74 @@
...
+                 class="text-center">
+                <div class="alert my-0 bg-100">
+                    <h6>
+                        <i class="fa fa-lock"/>
+                        <span>Course Locked</span>
+                    </h6>
+                    <div>
+                        <small>
+                            Finish
+                            <a t-if="len(channel.prerequisite_channel_ids) == 1"
```
Comment (awa-odoo) [inline addons/website_slides/views/website_slides_templates_course.xml:316]: Same question about opening in a tab for every links, maybe that is not what we want to do, to double-check with François.

File: addons/website_slides/static/src/js/slides_course_enroll_email.js [hunk 11]
```diff
@@ -58,7 +58,7 @@ publicWidget.registry.websiteSlidesEnroll = publicWidget.Widget.extend({
     _openDialog: function (channelId) {
         new SlideEnrollDialog(this, {
             channelId: channelId,
-            $element: this.$el
+            $element: this.$el.closest('.alert')
```
Comment (awa-odoo) [inline addons/website_slides/static/src/js/slides_course_enroll_email.js:61]: Mmmh, I don't really see how this is related ^^

**Review by awa-odoo — COMMENTED** [review]
Hello @pko-odoo 

Nice little feature, even if purely informative as discussed.
Left some comments, a bit of everything (naming, questions, style, ...).

Feel free to ask if anything is unclear.

Thanks & cheers!

File: addons/website_slides/models/slide_channel.py [hunk 12]
```diff
@@ -540,6 +551,33 @@ def _compute_action_rights(self):
                 channel.can_comment = user_karma >= channel.karma_slide_comment
                 channel.can_vote = user_karma >= channel.karma_slide_vote
 
+    @api.depends('visibility')
+    def _compute_prerequisite_channel_ids(self):
+        for channel in self:
+            if all(c.visibility != channel.visibility for c in channel.prerequisite_channel_ids):
+                channel.prerequisite_channel_ids = False
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/website_slides/models/slide_channel.py:564]: Still seems strange to wipe everything only if all visibilities don't match.
It's very specific, not sure we need to add code for that use case.
People can clean manually if they want.

File: addons/website_slides/static/src/xml/slide_course_prerequisite.xml [hunk 13]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates>
+    <t t-name="slide.course.prerequisite">
+        <ul class="m-0 ps-3">
+            <li t-foreach="channels" t-as="channel">
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/website_slides/static/src/xml/slide_course_prerequisite.xml:5]: I thought it was generic to all templates. Would still add one for safety based on the course id.

File: addons/website_slides/static/src/js/slides_course_enroll_email.js (hunk 11 above)
Comment (awa-odoo) [reply addons/website_slides/static/src/js/slides_course_enroll_email.js:61]: I'm guessing you re-tested the other use case(s) then? :slightly_smiling_face:

**Review by awa-odoo — COMMENTED** [review]
Hello @pko-odoo 

To answer your question on the task chatter, it feels sufficient for me to have the information displayed as you are about to delete.
(Note that if you delete from the list view, then you don't see it, not sure if it's an issue or not.)
To me it's not critical, it will just remove the prerequisite from the other course, which is fine as this is purely informational.
But it would be nice to confirm with François :slightly_smiling_face: 

I also replied to a couple comments but I did not re-read the diff as I don't think you are finished with the previous review?

Tell me if you need anything else.

Cheers!

File: addons/website_slides/static/src/js/slides_course_enroll_email.js (hunk 11 above)
Comment (pko-odoo) [reply addons/website_slides/static/src/js/slides_course_enroll_email.js:61]: Yes I did :wink:

File: addons/website_sale_slides/views/website_slides_templates.xml [hunk 14]
```diff
@@ -93,7 +93,7 @@
 <template name="Course Sidebar (infos, CTA)" id="course_join" inherit_id="website_slides.course_join">
     <!-- Channel main template: override button to join channel -->
     <xpath expr="//div[hasclass('o_wslides_js_course_join')]" position="inside">
-        <t t-if="(not channel.is_member or channel.can_publish) and channel.enroll == 'payment'">
+        <t t-elif="(not channel.is_member or channel.can_publish) and channel.enroll == 'payment'">
```
Comment (awa-odoo) [inline addons/website_sale_slides/views/website_slides_templates.xml:96]: Sorry, maybe my previous comment was not clear enough, but I would rather that we keep a simple "if" here that re-test everything needed.

Adding a t-elif "inside" a div is very much dependent on the base template, and it's a bit hard to follow when you get here as you have to go inside the other module to see in which case this block is shown.

File: addons/website_sale_slides/views/website_slides_templates.xml [hunk 15]
```diff
@@ -125,6 +125,23 @@
                 <span t-field="channel.product_id.list_price" t-options="{'widget': 'monetary', 'display_currency': product_info['currency_id']}"/>
             </div>
         </div>
+        <t t-if="channel.prerequisite_channel_ids and not channel.prerequisite_user_has_completed">
```
Comment (awa-odoo) [inline addons/website_sale_slides/views/website_slides_templates.xml:128]: Note that this template ("course_buy_course_button") is also used in "course_option_buy_course_now".

If I remember well it's an option in the customize menu of the website.

Would be nice to re-test that it's compatible with your changes here if not already done.

File: addons/website_slides/models/slide_channel.py [hunk 16]
```diff
@@ -359,6 +359,16 @@ def _get_default_enroll_msg(self):
     can_review = fields.Boolean('Can Review', compute='_compute_action_rights', compute_sudo=False)
     can_comment = fields.Boolean('Can Comment', compute='_compute_action_rights', compute_sudo=False)
     can_vote = fields.Boolean('Can Vote', compute='_compute_action_rights', compute_sudo=False)
+    # prerequisite settings
+    prerequisite_channel_ids = fields.Many2many(
+        'slide.channel', 'slide_channel_prerequisite_slide_channel_rel', 'channel_id', 'prerequisite_channel_id',
+        string='Prerequisites', help='Prerequisite courses to complete before accessing to this one.',
```
Comment (awa-odoo) [inline addons/website_slides/models/slide_channel.py:365]: "before accessing this one" ? "to" seems strange when reading.

File: addons/website_slides/models/slide_channel.py [hunk 17]
```diff
@@ -359,6 +359,16 @@ def _get_default_enroll_msg(self):
...
+        domain="[('id', '!=', id), ('visibility', '=', visibility), ('website_published', '=', website_published)]")
+    prerequisite_of_channel_ids = fields.Many2many(
+        'slide.channel', 'slide_channel_prerequisite_slide_channel_rel', 'prerequisite_channel_id', 'channel_id',
+        string='Prerequisite Of', help='Courses that has this course as prerequisite.')
```
Comment (awa-odoo) [inline addons/website_slides/models/slide_channel.py:369]: Courses that have* ?

File: addons/website_slides/static/src/js/slides_course_prerequisite.js [hunk 18]
```diff
@@ -0,0 +1,25 @@
+/** @odoo-module **/
+
+import { qweb } from 'web.core';
+import publicWidget from 'web.public.widget';
+
+publicWidget.registry.websiteSlidesCoursePrerequisite = publicWidget.Widget.extend({
+    selector: '.o_wslides_js_prerequisite_course',
+
+    start() {
+        return this._super(...arguments).then(async () => {
```
Comment (awa-odoo) [inline addons/website_slides/static/src/js/slides_course_prerequisite.js:10]: You could use the "await" syntax.

One indentation level gained :partying_face:

**Review by awa-odoo — COMMENTED** [review]
@pko-odoo Some last minute wording and quick remarks.

After that we should be good to launch.

Thanks & cheers!

**Comment by awa-odoo** [comment]
@robodoo r+
